cppcache/src/PdxHelper.cpp

```diff
@@ -47,92 +47,105 @@ PdxHelper::PdxHelper() {}
+  }
+}
+
+void PdxHelper::serializePdxSerializable(
+    const std::shared_ptr<PdxSerializable>& pdxObject, CacheImpl* cacheImpl,
+    std::shared_ptr<PdxTypeRegistry>& pdxTypeRegistry, DataOutput& output) {
   auto&& className = pdxObject->getClassName();
-  auto localPdxType = pdxTypeRegistry->getLocalPdxType(className);
-
-  if (localPdxType == nullptr) {
-    // need to grab type info, as fromdata is not called yet
-
-    PdxWriterWithTypeCollector ptc(output, className, pdxTypeRegistry);
-    pdxObject->toData(ptc);
-    auto nType = ptc.getPdxLocalType();
-
-    nType->InitializeType();
-    int32_t nTypeId = pdxTypeRegistry->getPDXIdForType(
-        className.c_str(), DataOutputInternal::getPool(output), nType, true);
-    nType->setTypeId(nTypeId);
-
-    ptc.endObjectWriting();
-    pdxTypeRegistry->addLocalPdxType(className, nType);
-    pdxTypeRegistry->addPdxType(nTypeId, nType);
-
-    if (cacheImpl != nullptr) {
-      uint8_t* stPos = const_cast<uint8_t*>(output.getBuffer()) +
-                       ptc.getStartPositionOffset();
-      int pdxLen = PdxHelper::readInt32(stPos);
-      cachePerfStats.incPdxSerialization(
-          pdxLen + 1 + 2 * 4);  // pdxLen + 93 DSID + len + typeID
-    }
 
-  } else  // we know locasl type, need to see preerved data
-  {
-    // if object got from server than create instance of RemoteWriter otherwise
-    // local writer.
-    // now always remotewriter as we have API Read/WriteUnreadFields
-    // so we don't know whether user has used those or not;; Can we do some
-    // trick here?
-
-    auto createPdxRemoteWriter = [&]() -> PdxRemoteWriter {
-      if (auto pd = pdxTypeRegistry->getPreserveData(pdxObject)) {
-        auto mergedPdxType = pdxTypeRegistry->getPdxType(pd->getMergedTypeId());
-        return PdxRemoteWriter(output, mergedPdxType, pd, pdxTypeRegistry);
-      } else {
-        return PdxRemoteWriter(output, className, pdxTypeRegistry);
-      }
-    };
+  if (!pdxTypeRegistry->getLocalPdxType(className)) {
+    // need to grab type info, as fromData is not called yet
+    PdxWriterWithTypeCollector writer(output, className, pdxTypeRegistry);
+    pdxObject->toData(writer);
 
-    PdxRemoteWriter prw = createPdxRemoteWriter();
+    auto nType = writer.getPdxLocalType();
+    registerPdxType(cacheImpl, className, nType, output);
 
-    pdxObject->toData(prw);
-    prw.endObjectWriting();
+    writer.endObjectWriting();
 
-    //[ToDo] need to write bytes for stats
-    if (cacheImpl != nullptr) {
-      uint8_t* stPos = const_cast<uint8_t*>(output.getBuffer()) +
-                       prw.getStartPositionOffset();
-      int pdxLen = PdxHelper::readInt32(stPos);
-      cachePerfStats.incPdxSerialization(
-          pdxLen + 1 + 2 * 4);  // pdxLen + 93 DSID + len + typeID
-    }
+    cacheImpl->getCachePerfStats().incPdxSerialization(
+        writer.getBytesWritten() + 1 + 2 * 4);  // pdxLen  93 DSID  len  typeID
```

Comment:
Can anyone explain these magic numbers? Delete the comment for being meaningless.